Fix #4201: Dynamic module loading

gzm0 opened the pull request.

gzm0 commented:
There are two things ATM I'm unsure about:

- Should we call this `loadDynamic` (or `dynamicLoad`) instead?
- How are we going to avoid indirection when trying to dynamically load a native dependency? (can we teach the optimizer that?)

gzm0 commented on ir/src/main/scala/org/scalajs/ir/Printers.scala:
```diff
@@ -335,6 +335,14 @@ object Printers {
+          print("::")
+          print(flags)
+          print(method)
+          printArgs(args)
```
Needs test.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/VarGen.scala:
```diff
@@ -132,15 +132,60 @@ private[emitter] final class VarGen(jsGen: JSGen, nameGen: NameGen,
   }
 
+  /** Call what is stored in the given global var via dynamic import. */
+  def callDynamicGlobalVar[T: Scope](field: String, scope: T, args: List[Tree])(
```
Not super happy about having to bake the call into here. Unsure what a better option is.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/modulesplitter/MaxModuleAnalyzer.scala:
```diff
@@ -97,28 +109,121 @@ private object MaxModuleAnalyzer {
+   *
+   *  Merging these tags into a single `ModuleID` is delegated to the
+   *  caller.
+   */
```
@sjrd PTAL at this. Is it convincing enough that the algorithm is correct?

gzm0 replied:
Yes, exactly. But isn't this exactly the example I'm making here with your `c` being "the class", your `a` being `d` and your `b` being `c`?

From the simplification we can see that `c` (your `b`) does not appear in the final grouping.

I'm not sure I know what to explain better or how based on your comment :-/

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/standard/LinkedClass.scala:
```diff
@@ -54,6 +54,7 @@ final class LinkedClass(
 
     val staticDependencies: Set[ClassName],
     val externalDependencies: Set[String],
+    val dynamicDependencies: Set[ClassName],
```
Group these three into a class?

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/modulesplitter/MaxModuleAnalyzer.scala:
```diff
@@ -97,28 +107,123 @@ private object MaxModuleAnalyzer {
+
+  /** Set of shortest, mutually prefix-free paths.
+   *
+   *  Would like to make this AnyVal, but that throws Erasure on 2.11 in an
+   *  infinite loop.
```
I don't think it would be beneficial anyway, since they are put in the values of maps, so they would be boxed there anyway.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/modulesplitter/MaxModuleAnalyzer.scala:
```diff
@@ -97,28 +107,123 @@ private object MaxModuleAnalyzer {
+   *  Would like to make this AnyVal, but that throws Erasure on 2.11 in an
+   *  infinite loop.
+   */
+  private final class Paths[H, T] private (
+      private val content: mutable.Map[H, Paths[T, T]]) {
```
While this recursive `Paths[H, T]` that contains `Paths[T, T]` is quite clever, I am not sure it holds its weight against two separate classes `HeadPaths` and `TailPaths` with hard-coded `ModuleID` and `ClassName`.
* The `ends` method could be implemented in a more efficient way by not creating a builder at each recursion step (the `TailPaths` class' equivalent of `ends` could take the `Builder[ClassName]` as parameter rather than creating a new one each time)
* `isEmpty` is only necessary for `TailPaths`

`put` would have to be duplicated, though.

I am not sure. Just putting this out there for consideration.

gzm0 replied:
I have tried this. I couldn't find a way of doing so without duplicating `put` and the non-trivial parts of `ends`. Basically, the little logic that distinguishes empty entries and missing entries feels non trivial with a typical interface to a `Map`. If you feel nevertheless this should be simplified, I can try harder.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/VarGen.scala:
```diff
@@ -132,15 +132,63 @@ private[emitter] final class VarGen(jsGen: JSGen, nameGen: NameGen,
     }
   }
 
+  /** Call what is stored in the given global var via dynamic import. */
```
It looks like this comment is not completely accurate anymore?

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala:
```diff
@@ -2112,6 +2121,13 @@ private[emitter] class FunctionEmitter(sjsGen: SJSGen) {
 
+        case ApplyDynamicImport(flags, className, method, args) =>
+          val call = withDynamicGlobalVar("s", (className, method.name)) { v =>
+            js.Apply(v, transformTypedArgs(method.name, args))
```
This translation is a bit too naive. It only works if all the `args` are *pure* expressions (side-effect-free is not even enough). If they're not, we need to make sure to always evaluate the `args` *before* performing the `ImportCall`. The semantics of `ApplyDynamicImport(..., method, args)` should be:

1. Evaluate all the `args` as `argsValues`
2. Asynchronously load the module and execute `method(...argsValues)`

sjrd reviewed (changes requested):
Overall this looks very nice. It looks simpler than I expected :)

sjrd commented:
> * Should we call this `loadDynamic` (or `dynamicLoad`) instead?

Naming is hard. In the doc you wrote that it's a "dynamic import boundary". Perhaps we should call it `dynamicImportBoundary`? (although that's a mouthful).

I wonder if there is precedent for this concept in other languages, that we could take inspiration from?

> * How are we going to avoid indirection when trying to dynamically load a native dependency? (can we teach the optimizer that?)

I don't think I understand the question. :s Do you have an example?

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/VarGen.scala:
```diff
@@ -132,15 +132,63 @@ private[emitter] final class VarGen(jsGen: JSGen, nameGen: NameGen,
+
+      val moduleTree = config.moduleKind match {
+        case ModuleKind.NoModule =>
+          throw new AssertionError("multiple modules in NoModule mode")
```
Well, but if we get here, that code would be incorrect. Getting here basically means that not all classes are in the same module, but we are in `NoModule` mode. (if the classes are in the same module, they previous branch would trigger).

gzm0 commented on compiler/src/main/scala/org/scalajs/nscplugin/PrepJSInterop.scala:
```diff
@@ -375,6 +375,49 @@ abstract class PrepJSInterop[G <: Global with Singleton](val global: G)
+                  DefDef(ctorSym, Block(gen.mkMethodCall(
+                      Super(clsSym, tpnme.EMPTY), ObjectClass.primaryConstructor, Nil, Nil))),
+                  // def apply(): Any = body
+                  DefDef(applySym, body.setType(ObjectTpe))))
```
Yep, seems so. Using `AnyTpe` remove the need for this.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/modulesplitter/MaxModuleAnalyzer.scala:
```diff
@@ -52,71 +107,123 @@ private object MaxModuleAnalyzer {
+   *  `c`, it is necessarily already loaded.
+   *
+   *  Once we have obtained this minimal set of paths, we use the last
+   *  elements of the paths to determine the final module
```
Yes. Updated the comment.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/modulesplitter/MinModuleAnalyzer.scala:
```diff
@@ -84,7 +93,22 @@ private object MinModuleAnalyzer {
+       * state).
+       */
+      classInfo.dynamicDependencies
+        .filter(!nodes.contains(_))
```
I've added a comment. PTAL.

gzm0 reviewed (commented):
Addressed some of the easier comments. Still TODO:

- Proper de-nesting in `FunctionEmitter`.
- Suggested additional test cases.

gzm0 commented:
> I don't think I understand the question. :s Do you have an example?

What I mean is the following. Say we have:

```scala
object Foo {
  @js.native
  @JSImport("foo.js", "bar")
  def bar(x: Int): Int = js.native
}

object MyMain {
  def main(args: Array[String]): Unit = {
    if (condition) {
      js.dynamicImport {
        Foo.bar(4)
      }
    }
  }
}
```

Ideally, for the dynamic import we should generate:

```javascript
import("foo.js").then(m => m.bar(4))
```

Without any additional thunk / Scala.js generated module.

I'd like to make sure that this is something we can build in the future with the current abstraction.

gzm0 commented:
Still TODO:
- Test cases
- Native optimization
- Name

gzm0 commented:
I believe I have addressed all the comments. Re Name: Webpack calls this dynamic imports, so I think we should just keep the name.

Feedback required from @sjrd:
- Can we optimize native calls as described in https://github.com/scala-js/scala-js/pull/4262#issuecomment-721201693 (at a later point).
- Changes are ready for second review.

gzm0 commented:
Please ignore the previous message. This is not ready for review yet. A lot of stuff fails with the local definitions.

gzm0 commented:
The current version of this PR fails with an IR checking error:

```
file:/home/tschlatt/20p/scala-js/test-suite/js/src/test/require-multi-modules/org/scalajs/testsuite/jsinterop/SJSDynamicImportTest.scala(119:15:CreateJSClass): Invalid CreateJSClass of top-level class org.scalajs.testsuite.jsinterop.SJSDynamicImportTest$$anon$7$Local$2
```

It is genuine: For some reason, the SJS compiler emits `Local` as an abstract JS type, and not a class with an empty capture list. I'm invstigating.

gzm0 commented:
Seems like local JS classes inside anonymous classes are broken :)

gzm0 commented on compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala:
```diff
@@ -4996,6 +5003,31 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
+              s"${args.size} at $pos")
+
+          args.head match {
+            case Block(Nil, Apply(fun @ Select(New(tpt), _), args)) =>
```
What we could do is allow the block to have other statements. That would not break the user-visible semantics but allow most general purpose transforms.

gzm0 commented:
Only open question: Will we be able to optimize like described in https://github.com/scala-js/scala-js/pull/4262#issuecomment-721201693? (I have a WIP branch: https://github.com/gzm0/scala-js/tree/optimize-loading).

gzm0 commented:
@sjrd I realize that my last comment might not have been clear :) This is ready for review.

sjrd commented:
Ah OK! I'll find some time to review it in the next 1-2 days.

sjrd commented on compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala:
```diff
@@ -5001,6 +5008,45 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
+              val clsSym = tpt.symbol
+              val ctor = fun.symbol
+
+              assert(clsSym.isSubClass(DynamicImportThunkClass),
+                  s"expected subclass of DynamicImportThunk, got: $clsSym")
+              assert(ctor.isPrimaryConstructor,
+                  s"expected primary constructor, got: $ctor")
```
Perhaps include the current `pos` in the error messages, to ease debugging if this pops up in the wild.

sjrd commented on compiler/src/main/scala/org/scalajs/nscplugin/PrepJSInterop.scala:
```diff
@@ -377,6 +377,52 @@ abstract class PrepJSInterop[G <: Global with Singleton](val global: G)
+
+          assert(currentOwner.isTerm, s"unexpected owner: $currentOwner at $pos")
+
+          val clsSym = currentOwner.newClass(tpnme.ANON_CLASS_NAME, pos)
```
Aren't we supposed to append a unique `$i` at this point? Perhaps not, I'm not sure.

gzm0 replied:
No. The trees that reach `PrepJSInterop` for anonymous classes do not have an `$i` yet.

sjrd commented on compiler/src/main/scala/org/scalajs/nscplugin/PrepJSInterop.scala:
```diff
@@ -377,6 +377,52 @@ abstract class PrepJSInterop[G <: Global with Singleton](val global: G)
+               * })
+               */
+              gen.mkMethodCall(Runtime_dynamicImport,
+                  List(tpeArg.tpe),
```
I don't suppose there is an overload that lets us reuse `tpeArg` directly, rather than extracting its `tpe` and then rewrapping it in a `TypeTree`?

sjrd reviewed (changes requested):
This looks very nice. I'm pleasantly surprised by the code in `PrepJSInterop`, which is much cleaner than I expected it to be. I only have fairly minor comments, except a few more critical things in the `FunctionEmitter`.

gzm0 commented:
I believe I have addressed all the comments. Let's see if the CI passes :)

@sjrd there is still my question in https://github.com/scala-js/scala-js/pull/4262#issuecomment-721201693. Do you think we can teach the optimizer (in a later PR) to do this? I just want to be sure there is no blocker to this that I missed, because an additional import hop might be prohibitive for some applications.

sjrd commented:
That's tricky. In theory, sure, nothing goes against that. I don't think the *optimizer* (as `OptimizerCore`) could do that, since it works in the frontend of the linker which doesn't have module info. But we could have a `ModuleSet` optimizer which *should* be able to do that kind of optimization.

IIUC, the situation would be:
* There is an `ApplyDynamicImport` to some method `m` in module `mod`.
* The module `mod` has exactly one static dependency `dep` that is not also a static dependency at the call site of the `ApplyDynamicImport`.

If those apply, then replace `ApplyDynamicImport` by `ImportCall(dep).then(depMembers => ...)`.

The problem though is the `...` This might contain quite a bit of code. If it's just the static method (somehow), then we can easily inline it, and replace references to `dep` so that point to `depMembers`. But if there are classes and other large portions of code in `mod`, we can't inline everything. We would have to pass `depMembers` as an additional parameter to everything in `mod`, as a constructor argument to classes, etc. And that would have an adversarial effect on performance, because we're now passing that `depMembers` dynamically, whereas the engine's resolution could have made it statically resolved.

If it's not too big and it really is just the static method itself, I think it's doable.

sjrd commented:
Not merging yet, until the optimization question is resolved. But from side it's good to go.

gzm0 commented:
I think the scope of what I think we should optimize is much narrower: I'm only concerned with *external* modules. I look at it that way: If an internal module has a dependency that we'd want to inline, then the module analyzer didn't do it's job properly (I realize that this is not the entire truth, but for simple cases is). Or another way of looking at this is that with `FewestModules`  analysis, we won't (modulo bugs) have trivial "just-hop one further" modules. However, this is not the case for native dependencies.

So in practice, I'd want the following to be optimized:

```scala
???.nativeMember                  // SelectJSNativeMember(...)
???.nativeMember(...)             // JSFunctionApply(SelectJSNativeMember(...), args)
nativeObject.member               // JSSelect(LoadJSModule(...), ...)
nativeObject.member(...)          // JSMethodApply(LoadJSModule(...), ..., args)
new NativeClass(...)              // JSNew(LoadJSConstructor(...), args)
```

IIUC the optimizer has all the information it needs to translate this into a `JSImportCall`. The tricky part are the arguments because we mustn't lift them across the loading boundary (the inverse case we have in `FunctionEmitter`). Maybe we can restrict it to trivial expressions and local references.

From a usage point of view, that would work the same than captures in `Future`s for example:

```scala
val deferredFoo = js.dynamicImport {
  nativeMember(foo())
}

val eagerFoo = {
  val f = foo()
  js.dynamicImport { nativeMember(f) }
}
```

It feels to me that such an optimization would be doable in the optimizer (and the pre-transform infrastructure might lend itself quite well to it). WDYT?

sjrd commented:
Ah, I see. Yes, I believe that this restricted goal can be done in the optimizer. Here is an outline of how I would see things going:

* In addition to `MethodID.shouldInline`, we could have `shouldInlineDynamicImport` that is true for the shapes that you identified (`SelectJSNativeMember`, etc.)
* When encountering an `ApplyDynamicImport`, if the target has `shouldInlineDynamicImport`, then evaluate the `args` in `PreTransform`s and get its body
* Generate the appropriate `JSImportCall` with a `.then`, and replace in the body the args and the reference to the external module.

After that, the `ApplyDynamicImport` node is gone, and the native class/object/member won't be reachable anymore.

gzm0 commented:
Yeah, that's pretty much what I had in mind. In this case it's good to go.

There is just one thing I still wanna do, run the test suite in a browser on this. (I'll send a PR updating TESTING afterwards). Then I'll merge.

SG?

sjrd commented:
Sounds good :)

gzm0 commented:
Ok, so testHtml for modules is broken for a multitude of reasons:
- It moves files, so relative import paths are broken.
- Module tests depend on Node.js modules.

I think we can nevertheless merge this and fix these things afterwards.

sjrd commented:
> I think we can nevertheless merge this and fix these things afterwards.

Yes, let's do that.
